### PR TITLE
G-API Initial code upload

### PR DESCRIPTION
**Pull request by dmatveev**
This is the initial source code upload for new OpenCV 4 module "Graph API":

> OpenCV Graph API (or G-API) is a new OpenCV module targeted to make
> regular image processing fast and portable. These two goals are
> achieved by introducing a new graph-based model of execution.
> 
> Traditionally OpenCV provided a lot of stand-alone image processing
> functions  (see modules `core` and `imgproc`). Many of that functions
> are well-optimized (e.g. vectorized for specific CPUs, parallel, etc)
> but still the out-of-box optimization scope has been limited to a
> single function only -- optimizing the whole algorithm built atop of that
> functions was a responsibility of a programmer.
> 
> OpenCV 3.0 introduced _Transparent API_ (or _T-API_) which allowed to
> offload OpenCV function calls transparently to OpenCL devices and save
> on Host/Device data transfers with cv::UMat -- and it was a great step
> forward. However, T-API is a dynamic API -- user code still remains
> unconstrained and OpenCL kernels are enqueued in arbitrary order, thus
> eliminating further pipeline-level optimization potential.
> 
> G-API brings implicit graph model to OpenCV 4.0. Graph model captures
> all operations and its data dependencies in a pipeline and so provides
> G-API framework with extra information to do pipeline-level
> optimizations.

This is an _integration_ branch -- it will be updated with new changes.
Please don't push to this branch directly.

Documentation and samples are WIP and will be pushed in a week or two.

### CODE REVIEW

**Comment by mshabunin** [comment]
* Please remove trailing whitespaces and multiple blank lines at EOF: http://pullrequest.opencv.org/buildbot/builders/precommit_docs/builds/17454/steps/whitespace%20opencv/logs/stdio
* Looks like we need to patch https://github.com/opencv/ade - alloc.cpp (add code like [this](https://github.com/opencv/opencv/blob/b39cd06249213220e802bb64260727711d9fc98c/modules/core/src/alloc.cpp#L45-L49)) to fix build on OSX

File: modules/gapi/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+    "${CMAKE_CURRENT_LIST_DIR}/include/opencv2/${name}/*.h"
+    "${CMAKE_CURRENT_LIST_DIR}/include/opencv2/${name}/util/*.hpp"
+    "${CMAKE_CURRENT_LIST_DIR}/include/opencv2/${name}/cpu/*.hpp"
+    "${CMAKE_CURRENT_LIST_DIR}/include/opencv2/${name}/fluid/*.hpp"
+    "${CMAKE_CURRENT_LIST_DIR}/include/opencv2/${name}/own/*.hpp"
+    )
+
+set(gapi_srcs
+    # Front-end part
+    src/api/gapi_priv.cpp
```
[the remark this answers is not in the document]
Comment (alalek) [reply modules/gapi/CMakeLists.txt]: Use `ocv_list_add_prefix(gapi_srcs "${CMAKE_CURRENT_LIST_DIR}/")` for that.

File: modules/gapi/CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,92 @@
+# FIXME: Remove CXX11 check after complete switch to OpenCV 4 branch
+# (CI, bundle, workloads, etc)
+if (NOT HAVE_CXX11 OR NOT TARGET ade)
```
[the remark this answers is not in the document]
Comment (dmatveev) [reply modules/gapi/CMakeLists.txt]: I think we better rewrite that code and drop aligned_union

**Comment by dmatveev** [comment]
Seems the latest problem with build on macOS is similar to https://github.com/martinmoene/gsl-lite/issues/63

File: modules/gapi/include/opencv2/gapi/gkernel.hpp [hunk 3]
```diff
@@ -380,7 +382,7 @@ namespace gapi {
         // Leading 0 helps to handle case when KK is an empty list (kernels<>()).
 
         int unused[] = { 0, (pkg.include<KK>(), 0)... };
-        (void) unused;
+        cv::util::suppress_unused_warning(unused);
```
Comment (mshabunin) [inline modules/gapi/include/opencv2/gapi/gkernel.hpp]: There is `CV_UNUSED` macro in the _core_ module.

Comment (dmatveev) [reply modules/gapi/include/opencv2/gapi/gkernel.hpp]: As G-API is being decoupled from OpenCV in parallel, we'll need to have our own

**Comment by dmatveev** [comment]
The latest status: only 1 warning remaining on Windows, will fix in a couple of hours.

ARMv7 test fails (and I don't have a clue why): http://pullrequest.opencv.org/buildbot/builders/precommit_armv7/builds/11175

**Comment by dmatveev** [comment]
@alalek @mshabunin @vpisarev feel lucky today :)

File: modules/gapi/include/opencv2/gapi/core.hpp [hunk 4]
```diff
@@ -0,0 +1,1559 @@
...
+            if (sz.width != 0 && sz.height != 0)
+            {
+                return in.withSize(sz);
+            }
+            else
+            {
+                GAPI_Assert(fx != 0. && fy != 0.);
+                return in.withSize
+                    (Size(static_cast<int>(std::round(in.size.width  * fx)),
+                          static_cast<int>(std::round(in.size.height * fy))));
```
[the remark this answers is not in the document]
Comment (dmatveev) [reply modules/gapi/include/opencv2/gapi/core.hpp]: Not sure what you mean

[the remark this answers is not in the document]
Comment (alalek) [reply modules/gapi/include/opencv2/gapi/core.hpp]: OpenCV uses own `cvRound()` (nearest-to-even) in `cv::resize()`.

[the remark this answers is not in the document]
Comment (dmatveev) [reply modules/gapi/include/opencv2/gapi/core.hpp]: ah ok. We do pass tests against cv::resize() already, so let's put this stuff into the backlog so address later (within the release timeframe).

**Comment by vpisarev** [comment]
let's merge it!

File: modules/gapi/test/internal/gapi_transactions_test.cpp [hunk 5]
```diff
@@ -0,0 +1,222 @@
+// This file is part of OpenCV project.
+// It is subject to the license terms in the LICENSE file found in the top-level directory
+// of this distribution and at http://opencv.org/license.html.
+//
+// Copyright (C) 2018 Intel Corporation
+
+
+#include <ade/graph.hpp>
+#include "test_precomp.hpp"
```
[the remark this answers is not in the document]
Comment (dmatveev) [reply modules/gapi/test/internal/gapi_transactions_test.cpp]: This and all below issues are now resolved

**Comment by alalek** [comment]
@dmatveev Could you please take a look on [valgrind issues](http://pullrequest.opencv.org/buildbot/builders/master_valgrind-lin64-debug/builds/10114)?

**Pull request merged**